File: CS/RichEditAPISample/CodeUtils/CodeResultViewer.cs
using System;
using System.CodeDom.Compiler;
using System.IO;
using System.Reflection;
using System.Windows.Forms;
using Microsoft.CSharp;
using System.Globalization;
using DevExpress.Spreadsheet;

namespace RichEditAPISample
{
    public abstract class RichEditExampleCodeEvaluator : ExampleCodeEvaluator
    {
        protected override string[] GetModuleAssemblies()
        {
            return new string[] { 
                AssemblyInfo.SRAssemblyRichEditCore,
                AssemblyInfo.SRAssemblyRichEdit,
                AssemblyInfo.SRAssemblyBars,
                AssemblyInfo.SRAssemblyEditors
            };
        }
        protected override string GetExampleClassName()
        {
            return "RichEditCodeResultViewer.ExampleItem";
        }
    }
    #region RichEditCSExampleCodeEvaluator
    public class RichEditCSExampleCodeEvaluator : RichEditExampleCodeEvaluator
    {

        protected override CodeDomProvider GetCodeDomProvider()
        {
            return new CSharpCodeProvider();
        }
        const string codeStart =
      "using System;\r\n" +
      "using DevExpress.XtraRichEdit;\r\n" +
      "using DevExpress.XtraRichEdit.API.Native;\r\n" +
      "using DevExpress.XtraRichEdit.Services;\r\n" +
      "using DevExpress.XtraRichEdit.Commands;\r\n" +
      "using DevExpress.XtraBars;\r\n" +
      "using System.Drawing;\r\n" +
      "using System.Windows.Forms;\r\n" +
      "using DevExpress.Utils;\r\n" +
      "using System.IO;\r\n" +
      "using System.Diagnostics;\r\n" +
      "using System.Xml;\r\n" +
      "using System.Data;\r\n" +
      "using System.Collections.Generic;\r\n" +
      "using System.Linq;\r\n" +
      "using System.Globalization;\r\n" +
      "using System.Reflection;\r\n" +
      "using System.ComponentModel;\r\n" +
      "namespace RichEditCodeResultViewer { \r\n" +
      "public class ExampleItem { \r\n" +
      "        public static void Process(RichEditControl richEditControl, BarButtonItem buttonCustomAction) { \r\n" +
      "                FieldInfo fItemClick = typeof(BarItem).GetField(\"itemClick\", BindingFlags.Static | BindingFlags.NonPublic);\r\n" +
      "                object objItemClick = fItemClick.GetValue(buttonCustomAction);\r\n" +
      "                PropertyInfo piItemClick = buttonCustomAction.GetType().GetProperty(\"Events\", BindingFlags.NonPublic | BindingFlags.Instance);\r\n" +
      "                EventHandlerList listItemClick = (EventHandlerList)piItemClick.GetValue(buttonCustomAction, null);\r\n" +
      "                listItemClick.RemoveHandler(objItemClick, listItemClick[objItemClick]);\r\n" +
      "\r\n" +
      "                FieldInfo fEventPaint = typeof(Control).GetField(\"EventPaint\", BindingFlags.Static | BindingFlags.NonPublic);\r\n" +
      "                object objEventPaint = fEventPaint.GetValue(richEditControl);\r\n" +
      "                PropertyInfo piEventPaint = richEditControl.GetType().GetProperty(\"Events\", BindingFlags.NonPublic | BindingFlags.Instance);\r\n" +
      "                EventHandlerList listEventPaint = (EventHandlerList)piEventPaint.GetValue(richEditControl, null);\r\n" +
      "                listEventPaint.RemoveHandler(objEventPaint, listEventPaint[objEventPaint]);\r\n" +
      "\r\n" +
      "                FieldInfo fEventMouseHover = typeof(Control).GetField(\"EventMouseMove\", BindingFlags.Static | BindingFlags.NonPublic);\r\n" +
      "                object objEventMouseHover = fEventMouseHover.GetValue(richEditControl);\r\n" +
      "                PropertyInfo piMouseHover = typeof(Control).GetProperty(\"Events\", BindingFlags.NonPublic | BindingFlags.Instance);\r\n" +
      "                EventHandlerList listEventMouseHover = (EventHandlerList)piMouseHover.GetValue(richEditControl, null);\r\n" +
      "                listEventMouseHover.RemoveHandler(objEventMouseHover, listEventMouseHover[objEventMouseHover]);\r\n" +
      "\r\n";

        const string mainMethodEnd =
        "       \r\n }\r\n";
        
        const string codeEnd =
            "    }\r\n" +
            "}\r\n";
        protected override string CodeStart { get { return codeStart; } }
        protected override string MainMethodEnd { get { return mainMethodEnd; } }
        protected override string CodeEnd { get { return codeEnd; } }
    }
    #endregion
    #region RichEditVbExampleCodeEvaluator
    public class RichEditVbExampleCodeEvaluator : RichEditExampleCodeEvaluator
    {

        protected override CodeDomProvider GetCodeDomProvider()
        {
            return new Microsoft.VisualBasic.VBCodeProvider();
        }
        const string codeStart =
      "Imports Microsoft.VisualBasic\r\n" +
      "Imports System\r\n" +
      "Imports DevExpress.XtraRichEdit\r\n" +
      "Imports DevExpress.XtraRichEdit.API.Native\r\n" +
      "Imports DevExpress.XtraRichEdit.Services\r\n" +
      "Imports DevExpress.XtraRichEdit.Commands\r\n" +
      "Imports DevExpress.XtraBars\r\n" +
      "Imports System.Drawing\r\n" +
      "Imports System.Windows.Forms\r\n" +
      "Imports DevExpress.Utils\r\n" +
      "Imports System.IO\r\n" +
      "Imports System.Diagnostics\r\n" +
      "Imports System.Xml\r\n" +
      "Imports System.Data\r\n" +
      "Imports System.Collections.Generic\r\n" +
      "Imports System.Linq;\r\n" +
      "Imports System.Globalization\r\n" +
      "Imports System.Reflection;\r\n" +
      "Imports System.ComponentModel;\r\n" +
      "Namespace RichEditCodeResultViewer\r\n" +
      "	Public Class ExampleItem\r\n" +
      "		Public Shared Sub Process(ByVal richEditControl As RichEditControl, ByVal buttonCustomAction As BarButtonItem)\r\n" +
      "                Dim fItemClick As FieldInfo = GetType(BarItem).GetField(\"itemClick\", BindingFlags.Static Or BindingFlags.NonPublic)\r\n" +
      "                Dim objItemClick As Object = fItemClick.GetValue(buttonCustomAction)\r\n" +
      "                Dim piItemClick As PropertyInfo = buttonCustomAction.GetType().GetProperty(\"Events\", BindingFlags.NonPublic Or BindingFlags.Instance)\r\n" +
      "                Dim listItemClick As EventHandlerList = CType(piItemClick.GetValue(buttonCustomAction, Nothing), EventHandlerList)\r\n" +
      "                listItemClick.RemoveHandler(objItemClick, listItemClick(objItemClick))\r\n" +
      "\r\n" +
      "                Dim fEventPaint As FieldInfo = GetType(Control).GetField(\"EventPaint\", BindingFlags.Static Or BindingFlags.NonPublic)\r\n" +
      "                Dim objEventPaint As Object = fEventPaint.GetValue(richEditControl)\r\n" +
      "                Dim piEventPaint As PropertyInfo = richEditControl.GetType().GetProperty(\"Events\", BindingFlags.NonPublic Or BindingFlags.Instance)\r\n" +
      "                Dim listEventPaint As EventHandlerList = CType(piEventPaint.GetValue(richEditControl, Nothing), EventHandlerList)\r\n" +
      "                listEventPaint.RemoveHandler(objEventPaint, listEventPaint(objEventPaint))\r\n" +
      "\r\n" +
      "                Dim fEventMouseHover As FieldInfo = GetType(Control).GetField(\"EventMouseMove\", BindingFlags.Static Or BindingFlags.NonPublic)\r\n" +
      "                Dim objEventMouseHover As Object = fEventMouseHover.GetValue(richEditControl)\r\n" +
      "                Dim piMouseHover As PropertyInfo = GetType(Control).GetProperty(\"Events\", BindingFlags.NonPublic Or BindingFlags.Instance)\r\n" +
      "                Dim listEventMouseHover As EventHandlerList = CType(piMouseHover.GetValue(richEditControl, Nothing), EventHandlerList)\r\n" +
      "                listEventMouseHover.RemoveHandler(objEventMouseHover, listEventMouseHover(objEventMouseHover))\r\n" +
      "\r\n";

        const string mainMethodEnd =
        "\r\n		End Sub\r\n\r\n";

        const string codeEnd = 
        "	End Class\r\n" +
        "End Namespace\r\n";

        protected override string CodeStart { get { return codeStart; } }
        protected override string MainMethodEnd { get { return mainMethodEnd; } }
        protected override string CodeEnd { get { return codeEnd; } }
    }
    #endregion
}

File: VB/RichEditAPISample/CodeUtils/CodeResultViewer.vb
Imports Microsoft.VisualBasic
Imports System
Imports System.CodeDom.Compiler
Imports System.IO
Imports System.Reflection
Imports System.Windows.Forms
Imports Microsoft.CSharp
Imports System.Globalization
Imports DevExpress.Spreadsheet

Namespace RichEditAPISample
	Public MustInherit Class RichEditExampleCodeEvaluator
		Inherits ExampleCodeEvaluator
		Protected Overrides Function GetModuleAssemblies() As String()
			Return New String() { AssemblyInfo.SRAssemblyRichEditCore, AssemblyInfo.SRAssemblyRichEdit, AssemblyInfo.SRAssemblyBars, AssemblyInfo.SRAssemblyEditors }
		End Function
		Protected Overrides Function GetExampleClassName() As String
			Return "RichEditCodeResultViewer.ExampleItem"
		End Function
	End Class
	#Region "RichEditCSExampleCodeEvaluator"
	Public Class RichEditCSExampleCodeEvaluator
		Inherits RichEditExampleCodeEvaluator

		Protected Overrides Function GetCodeDomProvider() As CodeDomProvider
			Return New CSharpCodeProvider()
		End Function
        Private Const codeStart_Renamed As String = "using System;" & Constants.vbCrLf & "using DevExpress.XtraRichEdit;" & Constants.vbCrLf & "using DevExpress.XtraRichEdit.API.Native;" & Constants.vbCrLf & "using DevExpress.XtraRichEdit.Services;" & Constants.vbCrLf & "using DevExpress.XtraRichEdit.Commands;" & Constants.vbCrLf & "using DevExpress.XtraBars;" & Constants.vbCrLf & "using System.Drawing;" & Constants.vbCrLf & "using System.Windows.Forms;" & Constants.vbCrLf & "using DevExpress.Utils;" & Constants.vbCrLf & "using System.IO;" & Constants.vbCrLf & "using System.Diagnostics;" & Constants.vbCrLf & "using System.Xml;" & Constants.vbCrLf & "using System.Data;" & Constants.vbCrLf & "using System.Collections.Generic;" & Constants.vbCrLf & "using System.Linq;" & Constants.vbCrLf & "using System.Globalization;" & Constants.vbCrLf & "using System.Reflection;" & Constants.vbCrLf & "using System.ComponentModel;" & Constants.vbCrLf & "namespace RichEditCodeResultViewer { " & Constants.vbCrLf & "public class ExampleItem { " & Constants.vbCrLf & "        public static void Process(RichEditControl richEditControl, BarButtonItem buttonCustomAction) { " & Constants.vbCrLf & "                FieldInfo fItemClick = typeof(BarItem).GetField(""itemClick"", BindingFlags.Static | BindingFlags.NonPublic);" & Constants.vbCrLf & "                object objItemClick = fItemClick.GetValue(buttonCustomAction);" & Constants.vbCrLf & "                PropertyInfo piItemClick = buttonCustomAction.GetType().GetProperty(""Events"", BindingFlags.NonPublic | BindingFlags.Instance);" & Constants.vbCrLf & "                EventHandlerList listItemClick = (EventHandlerList)piItemClick.GetValue(buttonCustomAction, null);" & Constants.vbCrLf & "                listItemClick.RemoveHandler(objItemClick, listItemClick[objItemClick]);" & Constants.vbCrLf & Constants.vbCrLf & "                FieldInfo fEventPaint = typeof(Control).GetField(""EventPaint"", BindingFlags.Static | BindingFlags.NonPublic);" & Constants.vbCrLf & "                object objEventPaint = fEventPaint.GetValue(richEditControl);" & Constants.vbCrLf & "                PropertyInfo piEventPaint = richEditControl.GetType().GetProperty(""Events"", BindingFlags.NonPublic | BindingFlags.Instance);" & Constants.vbCrLf & "                EventHandlerList listEventPaint = (EventHandlerList)piEventPaint.GetValue(richEditControl, null);" & Constants.vbCrLf & "                listEventPaint.RemoveHandler(objEventPaint, listEventPaint[objEventPaint]);" & Constants.vbCrLf & Constants.vbCrLf & "                FieldInfo fEventMouseHover = typeof(Control).GetField(""EventMouseMove"", BindingFlags.Static | BindingFlags.NonPublic);" & Constants.vbCrLf & "                object objEventMouseHover = fEventMouseHover.GetValue(richEditControl);" & Constants.vbCrLf & "                PropertyInfo piMouseHover = typeof(Control).GetProperty(""Events"", BindingFlags.NonPublic | BindingFlags.Instance);" & Constants.vbCrLf & "                EventHandlerList listEventMouseHover = (EventHandlerList)piMouseHover.GetValue(richEditControl, null);" & Constants.vbCrLf & "                listEventMouseHover.RemoveHandler(objEventMouseHover, listEventMouseHover[objEventMouseHover]);" & Constants.vbCrLf & Constants.vbCrLf

		Private Const mainMethodEnd_Renamed As String = "       " & Constants.vbCrLf & " }" & Constants.vbCrLf

		Private Const codeEnd_Renamed As String = "    }" & Constants.vbCrLf & "}" & Constants.vbCrLf
		Protected Overrides ReadOnly Property CodeStart() As String
			Get
				Return codeStart_Renamed
			End Get
		End Property
		Protected Overrides ReadOnly Property MainMethodEnd() As String
			Get
				Return mainMethodEnd_Renamed
			End Get
		End Property
		Protected Overrides ReadOnly Property CodeEnd() As String
			Get
				Return codeEnd_Renamed
			End Get
		End Property
	End Class
	#End Region
	#Region "RichEditVbExampleCodeEvaluator"
	Public Class RichEditVbExampleCodeEvaluator
		Inherits RichEditExampleCodeEvaluator

		Protected Overrides Function GetCodeDomProvider() As CodeDomProvider
			Return New Microsoft.VisualBasic.VBCodeProvider()
		End Function
        Private Const codeStart_Renamed As String = "Imports Microsoft.VisualBasic" & Constants.vbCrLf & "Imports System" & Constants.vbCrLf & "Imports DevExpress.XtraRichEdit" & Constants.vbCrLf & "Imports DevExpress.XtraRichEdit.API.Native" & Constants.vbCrLf & "Imports DevExpress.XtraRichEdit.Services" & Constants.vbCrLf & "Imports DevExpress.XtraRichEdit.Commands" & Constants.vbCrLf & "Imports DevExpress.XtraBars" & Constants.vbCrLf & "Imports System.Drawing" & Constants.vbCrLf & "Imports System.Windows.Forms" & Constants.vbCrLf & "Imports DevExpress.Utils" & Constants.vbCrLf & "Imports System.IO" & Constants.vbCrLf & "Imports System.Diagnostics" & Constants.vbCrLf & "Imports System.Xml" & Constants.vbCrLf & "Imports System.Data" & Constants.vbCrLf & "Imports System.Collections.Generic" & Constants.vbCrLf & "Imports System.Linq" & Constants.vbCrLf & "Imports System.Globalization" & Constants.vbCrLf & "Imports System.Reflection" & Constants.vbCrLf & "Imports System.ComponentModel" & Constants.vbCrLf & "Namespace RichEditCodeResultViewer" & Constants.vbCrLf & "	Public Class ExampleItem" & Constants.vbCrLf & "		Public Shared Sub Process(ByVal richEditControl As RichEditControl, ByVal buttonCustomAction As BarButtonItem)" & Constants.vbCrLf & "                Dim fItemClick As FieldInfo = GetType(BarItem).GetField(""itemClick"", BindingFlags.Static Or BindingFlags.NonPublic)" & Constants.vbCrLf & "                Dim objItemClick As Object = fItemClick.GetValue(buttonCustomAction)" & Constants.vbCrLf & "                Dim piItemClick As PropertyInfo = buttonCustomAction.GetType().GetProperty(""Events"", BindingFlags.NonPublic Or BindingFlags.Instance)" & Constants.vbCrLf & "                Dim listItemClick As EventHandlerList = CType(piItemClick.GetValue(buttonCustomAction, Nothing), EventHandlerList)" & Constants.vbCrLf & "                listItemClick.RemoveHandler(objItemClick, listItemClick(objItemClick))" & Constants.vbCrLf & Constants.vbCrLf & "                Dim fEventPaint As FieldInfo = GetType(Control).GetField(""EventPaint"", BindingFlags.Static Or BindingFlags.NonPublic)" & Constants.vbCrLf & "                Dim objEventPaint As Object = fEventPaint.GetValue(richEditControl)" & Constants.vbCrLf & "                Dim piEventPaint As PropertyInfo = richEditControl.GetType().GetProperty(""Events"", BindingFlags.NonPublic Or BindingFlags.Instance)" & Constants.vbCrLf & "                Dim listEventPaint As EventHandlerList = CType(piEventPaint.GetValue(richEditControl, Nothing), EventHandlerList)" & Constants.vbCrLf & "                listEventPaint.RemoveHandler(objEventPaint, listEventPaint(objEventPaint))" & Constants.vbCrLf & Constants.vbCrLf & "                Dim fEventMouseHover As FieldInfo = GetType(Control).GetField(""EventMouseMove"", BindingFlags.Static Or BindingFlags.NonPublic)" & Constants.vbCrLf & "                Dim objEventMouseHover As Object = fEventMouseHover.GetValue(richEditControl)" & Constants.vbCrLf & "                Dim piMouseHover As PropertyInfo = GetType(Control).GetProperty(""Events"", BindingFlags.NonPublic Or BindingFlags.Instance)" & Constants.vbCrLf & "                Dim listEventMouseHover As EventHandlerList = CType(piMouseHover.GetValue(richEditControl, Nothing), EventHandlerList)" & Constants.vbCrLf & "                listEventMouseHover.RemoveHandler(objEventMouseHover, listEventMouseHover(objEventMouseHover))" & Constants.vbCrLf & Constants.vbCrLf

        Private Const mainMethodEnd_Renamed As String = Constants.vbCrLf & "		End Sub" & Constants.vbCrLf

        Private Const codeEnd_Renamed As String = Constants.vbCrLf & "	End Class" & Constants.vbCrLf & "End Namespace" & Constants.vbCrLf

		Protected Overrides ReadOnly Property CodeStart() As String
			Get
				Return codeStart_Renamed
			End Get
		End Property
		Protected Overrides ReadOnly Property MainMethodEnd() As String
			Get
				Return mainMethodEnd_Renamed
			End Get
		End Property
		Protected Overrides ReadOnly Property CodeEnd() As String
			Get
				Return codeEnd_Renamed
			End Get
		End Property
	End Class
	#End Region
End Namespace
